File: data_structure/radix_sort.py
def radix_sort(nums):
    m = max(nums)
    n = len(nums)
    step = 1
    while step < m:
        keys = [(a//step)%10 for a in nums]
        counting_sort(keys, nums)
        print(nums)
        step *= 10

def counting_sort(keys, nums):
    n = len(keys)
    C = [0 for _ in range(11)] # O(k) extra space, k = 𝚯(n)
    B = [0 for _ in range(n)] # O(n) extra space
    for a in keys:
        C[a] += 1
    for i in range(1, len(C)):
        C[i] = C[i] + C[i-1]
    for i in range(n-1, -1, -1):
        B[C[keys[i]]-1] = nums[i]
        C[keys[i]] -= 1
    for i in range(len(nums)):
        nums[i] = B[i]

radix_sort([109, 209, 301, 333, 102, 103, 55, 3])

File: solutions/0164_Maximum_Gap/radix_sort.java
class Solution {
    public int maximumGap(int[] nums) {
        if (nums.length <= 1) return 0;
        radixSort(nums);
        int best = 0;
        for (int i = 0; i < nums.length - 1; ++i) {
            best = Math.max(best, nums[i+1] - nums[i]);
        }
        return best;
    }

    private void radixSort(int[] nums) {
        int[] keys = new int[nums.length];
        int m = 0, n = nums.length;
        int step = 1;
        for (int i = 0; i < nums.length; ++i) {
            m = Math.max(m, nums[i]);
        }
        while (step <= m) {
            for (int i = 0; i < n; ++i) keys[i] = (nums[i] / step) % 10;
            countSort(keys, nums);
            step *= 10;
        }
    }

    private void countSort(int[] keys, int[] nums) {
        int n = keys.length;
        int[] C = new int[11];
        Arrays.fill(C, 0);
        int[] B = new int[n];
        for (int a : keys) C[a] += 1;
        for (int i = 1; i < 11; ++i) C[i] = C[i] + C[i-1];
        for (int i = n-1; i >= 0; --i) {
            B[C[keys[i]]] = nums[i];
            C[keys[i]] -= 1;
        }
        for (int i = 0; i < n; ++i) nums[i] = B[i];
    }
}
